core/resource/resource.go

```diff
@@ -0,0 +1,309 @@
+
+	tenant tenant.Tenant
+
+	spec     map[string]any
+	metadata *Metadata
+
+	status Status
+}
+
+func NewResource(fullName string, kind Kind, store Store, tnnt tenant.Tenant, meta *Metadata, spec map[string]any) (*Resource, error) {
+	sections := strings.Split(fullName, nameSectionSeparator)
+	var strName string
+	if kind == KindDataset {
+		if len(sections) != DatesetNameSections {
+			return nil, errors.InvalidArgument(EntityResource, "invalid dataset name: "+fullName)
+		}
+		strName = sections[1]
+	} else {
+		if len(sections) != TableNameSections {
+			return nil, errors.InvalidArgument(EntityResource, "invalid resource name: "+fullName)
+		}
+		strName = sections[2]
+	}
+
+	name, err := NameFrom(strName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(spec) == 0 {
+		return nil, errors.InvalidArgument(EntityResource, "invalid resource spec for "+fullName)
+	}
+
+	dataset, err := DataSetFrom(store, sections[0], sections[1])
+	if err != nil {
+		return nil, err
+	}
+
+	if meta == nil {
+		return nil, errors.InvalidArgument(EntityResource, "invalid resource metadata")
+	}
+
+	return &Resource{
+		name:     name,
+		kind:     kind,
+		dataset:  dataset,
+		tenant:   tnnt,
+		spec:     spec,
+		metadata: meta,
+		status:   StatusUnknown,
+	}, nil
+}
+
+func (r *Resource) Name() Name {
+	return r.name
+}
+
+func (r *Resource) FullName() string {
+	if r.kind == KindDataset {
+		return r.dataset.FullName()
+	}
+	return r.dataset.FullName() + "." + r.name.String()
+}
+
+func (r *Resource) URN() string {
+	if r.kind == KindDataset {
+		return r.dataset.URN()
+	}
+	return r.dataset.URN() + "." + r.name.String()
+}
+
+func (r *Resource) Metadata() *Metadata {
+	return r.metadata
+}
+
+func (r *Resource) Kind() Kind {
+	return r.kind
+}
+
+func (r *Resource) Tenant() tenant.Tenant {
+	return r.tenant
+}
+
+func (r *Resource) Dataset() Dataset {
+	return r.dataset
+}
+
+func (r *Resource) Status() Status {
+	return r.status
+}
+
+func (r *Resource) Spec() map[string]any {
+	return r.spec
+}
+
+func (r *Resource) Validate() error {
+	switch r.kind {
+	case KindTable:
+		table, err := ConvertSpecTo[Table](r)
+		if err != nil {
+			return err
+		}
+		table.Name = r.name
+		table.Dataset = r.dataset
+		return table.Validate()
+
+	case KindExternalTable:
+		externalTable, err := ConvertSpecTo[ExternalTable](r)
+		if err != nil {
+			return err
+		}
+		externalTable.Name = r.name
+		externalTable.Dataset = r.dataset
+		return externalTable.Validate()
+
+	case KindView:
+		view, err := ConvertSpecTo[View](r)
+		if err != nil {
+			return err
+		}
+		view.Name = r.name
+		view.Dataset = r.dataset
+		return view.Validate()
+
+	case KindDataset:
+		ds, err := ConvertSpecTo[DatasetDetails](r)
+		if err != nil {
+			return err
+		}
+		ds.Dataset = r.dataset
+		return ds.Validate()
+
+	default:
+		return errors.InvalidArgument(EntityResource, "unknown kind")
+	}
+}
+
+func (r *Resource) Equal(incoming *Resource) bool {
+	if r == nil || incoming == nil {
```

Comment:
`r` cannot be nil, as it is a method call, we will have nil pointer error. So we can omit that condition.

Reply:
Technically, `r` can be nil. If `r` is nil, we will encounter panic if we remove this condition. Now, if we assume that `r` cannot be nil because of how we designed Optimus, then we may as well assume that `incoming` cannot be nil. If that's the case, then we can just remove `nil` check condition, such as in line `187` mentioned here.